src/dsp_tools/commands/xmlupload/xmlupload.py

```diff
@@ -71,59 +79,71 @@ def xmlupload(
         uploaded because there is an error in it
     """
 
+    default_ontology, root, shortcode = validate_and_parse_xml_file(input_file=input_file, imgdir=imgdir)
+
     con = ConnectionLive(creds.server)
     con.login(creds.user, creds.password)
+    config = UploadConfig(interrupt_after=interrupt_after).with_server_info(server=creds.server, shortcode=shortcode)
 
-    default_ontology, root, shortcode = validate_and_parse_xml_file(
-        input_file=input_file,
-        imgdir=imgdir,
-        preprocessing_done=config.media_previously_uploaded,
-    )
+    ontology_client = OntologyClientLive(con=con, shortcode=shortcode, default_ontology=default_ontology)
+    resources, permissions_lookup, stash = prepare_upload(root, ontology_client)
 
-    config = config.with_server_info(
-        server=creds.server,
-        shortcode=shortcode,
-    )
+    clients = _get_live_clients(con, creds, shortcode, imgdir)
+    state = UploadState(resources, stash, config, permissions_lookup)
 
-    ingest_client: AssetClient
-    if config.media_previously_uploaded:
-        ingest_client = BulkIngestedAssetClient()
-    else:
-        ingest_client = DspIngestClientLive(
-            dsp_ingest_url=creds.dsp_ingest_url,
-            token=con.get_token(),
-            shortcode=config.shortcode,
-            imgdir=imgdir,
-        )
+    return execute_upload(clients, state)
 
-    ontology_client = OntologyClientLive(
-        con=con,
-        shortcode=shortcode,
-        default_ontology=default_ontology,
-    )
-    do_xml_consistency_check_with_ontology(onto_client=ontology_client, root=root)
 
-    resources, permissions_lookup, stash = _prepare_upload(
-        root=root,
-        con=con,
-        default_ontology=default_ontology,
+def _get_live_clients(
+    con: Connection,
+    creds: ServerCredentials,
+    shortcode: str,
+    imgdir: str,
+) -> UploadClients:
+    ingest_client: AssetClient
+    ingest_client = DspIngestClientLive(
+        dsp_ingest_url=creds.dsp_ingest_url,
+        token=con.get_token(),
+        shortcode=shortcode,
+        imgdir=imgdir,
     )
-
-    project_client: ProjectClient = ProjectClientLive(con, config.shortcode)
+    project_client: ProjectClient = ProjectClientLive(con, shortcode)
     list_client: ListClient = ListClientLive(con, project_client.get_project_iri())
-    upload_state = UploadState(resources, [], IriResolver(), stash, config, permissions_lookup)
-
-    upload_resources(
-        upload_state=upload_state,
-        ingest_client=ingest_client,
+    return UploadClients(
+        asset_client=ingest_client,
         project_client=project_client,
         list_client=list_client,
     )
 
-    return cleanup_upload(upload_state)
+
+def execute_upload(clients: UploadClients, upload_state: UploadState) -> bool:
+    """Execute an upload from an upload state, and clean up afterwards.
+
+    Args:
+        clients: the clients needed for the upload (AssetClient, ProjectClient, ListClient)
```

Comment:
this will soon fall prey to docs rotting. If someone modifies the fields of the `UploadClients` class, they will forget to update this docstring